### PR TITLE
BUG: Fix insertion of wrong-dtypes NaT into Series[m8ns]

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken off from #27311 to troubleshoot build-specific failures.  Also re-wrote test to parametrize and be more succinct

### CODE REVIEW

File: pandas/tests/series/indexing/test_indexing.py [hunk 1]
```diff
@@ -654,6 +654,37 @@ def test_timedelta_assignment():
     tm.assert_series_equal(s, expected)
 
 
+class TestNaTAssignment:
```
Comment (simonjayhawkins) [inline pandas/tests/series/indexing/test_indexing.py]: do we need a class here?

Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: This is broken off from a PR that implemented several related tests.

File: pandas/_libs/index.pyx [hunk 2]
```diff
@@ -544,6 +544,10 @@ cpdef convert_scalar(ndarray arr, object value):
             pass
         elif isinstance(value, timedelta):
             return Timedelta(value).value
+        elif util.is_datetime64_object(value):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: how does this not hit the if at line 531?

**Review by jreback — CHANGES_REQUESTED** [review]
not sure I agree with the point of this PR. the np.* based 'NaT' should *never* exist in pandas at all, ever, we simply need to cast them appropriately; I get that you want to turn an M8 array when presented with an m8('NaT') into object, but I do't think we currently do. However I don't view this distinct as very import and it adds a *lot* of complexity. We now treat NaT, np.nan, None and np.*('nat') synonymously as missing values when presented to say a M8 (or a m8); I don't see why we don't  simply say, hey I have a missig value, great use the value for this dtype.

What you are trying to do here IMHO is inserting *even* more of numpy issues into pandas when we are trying to do the opposite.

File: pandas/_libs/index.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/index.pyx]: 531 is a check on the `arr` variable.  This is a check on `value`.

**Comment by jbrockmendel** [comment]
> not sure I agree with the point of this PR. the np.* based 'NaT' should never exist in pandas at all, ever, we simply need to cast them appropriately

xref #24983.  What you are suggesting is 100% impossible to do correctly.

> I get that you want to turn an M8 array when presented with an m8('NaT') into object, but I do't think we currently do

Well you're right that it is not what we currently do, but that is a bug.  It is unambiguously the correct behavior.

> However I don't view this distinct as very import and it adds a lot of complexity

It's your prerogative (a word whose spelling never ceases to amaze me) to consider it a low-priority bug.  On the flip side, I see complexity stemming from the fact that we have to guess what dtype NaT is behaving as in any given situation, and the impossibility of always guessing correctly.

> We now treat NaT, np.nan, None and np.*('nat') synonymously as missing values when presented to say a M8 (or a m8); I don't see why we don't simply say, hey I have a missig value, great use the value for this dtype.

Consider: if a user passes `pd.NaT`, that does basically your "hey I have a missing value" thing.  But if a user _specifically_ passes `np.datetime64("NaT")`, they did that on purpose.

File: pandas/_libs/index.pyx [hunk 3]
```diff
@@ -544,6 +544,10 @@ cpdef convert_scalar(ndarray arr, object value):
...
+            # exclude np.datetime64("NaT") which would otherwise be picked up
+            #  by the `value != value check below
+            pass
```
Comment (jbrockmendel) [inline pandas/_libs/index.pyx]: really everything from 531 down to 555 should be ripped out and this handled by the DatetimeArray/TimedeltaArray `__setitem__` implementation

**Comment by jbrockmendel** [comment]
changed test class to just a function

**Comment by jbrockmendel** [comment]
rebased following can_hold_element fix

**Comment by jbrockmendel** [comment]
gentle ping; this one is a bugfix, admittedly a small one

**Review by TomAugspurger — APPROVED** [review]
LGTM. @jreback?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
sure

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**